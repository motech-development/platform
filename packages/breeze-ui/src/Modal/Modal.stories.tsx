import { boolean, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import BaseStyles from '../BaseStyles/BaseStyles';
import Button from '../Button/Button';
import Typography from '../Typography/Typography';
import Modal from './Modal';

const stories = storiesOf('Modal', module);
const onDismiss = () => {};

stories.addDecorator(withKnobs);

stories.add('Basic modal', () => (
  <>
    <BaseStyles />

    <Modal isOpen={boolean('Show', true)} onDismiss={onDismiss}>
      <Typography rule component="h1" variant="h2" margin="lg">
        Are you sure you want to do this?
      </Typography>
      <Typography component="p" variant="p" margin="lg">
        Clicking OK will delete this permanently
      </Typography>
      <Button colour="danger">Delete</Button>{' '}
      <Button>Actually, don&#39;t bother</Button>
    </Modal>
  </>
));
