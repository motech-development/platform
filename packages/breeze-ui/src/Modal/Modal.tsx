import { FC, MouseEvent, ReactNode, Suspense, useEffect } from 'react';
import { createPortal } from 'react-dom';
import styled from 'styled-components';
import Card from '../Card/Card';
import Loader from '../Loader/Loader';
import Overlay from '../Overlay/Overlay';

type TModalSize = 'sm' | 'lg';

const ModalOuter = styled.div`
  bottom: 0;
  left: 0;
  outline: 0;
  overflow-x: hidden;
  overflow-y: auto;
  position: fixed;
  right: 0;
  top: 0;
  z-index: 1100;
`;

const ModalContent = styled.div`
  outline: 0;
  position: relative;
  width: 100%;
  pointer-events: auto;
`;

interface IModalDialogProps {
  size: TModalSize;
}

const ModalDialog = styled.div<IModalDialogProps>`
  box-shadow:
    0 3px 5px -1px rgba(0, 0, 0, 0.2),
    0 5px 8px 0 rgba(0, 0, 0, 0.14),
    0 1px 14px 0 rgba(0, 0, 0, 0.12);
  margin: 5rem auto;
  max-width: ${({ size }) => (size === 'sm' ? '500px' : '90vw')};
  pointer-events: none;
  position: relative;
`;

export interface IModalProps {
  children: ReactNode;
  isOpen: boolean;
  size?: TModalSize;
  title: string;
  onDismiss: () => void;
}

const Modal: FC<IModalProps> = ({
  children,
  isOpen,
  onDismiss,
  size = 'sm',
  title,
}) => {
  const output = document.createElement('div');
  const doNotDismiss = (e: MouseEvent<HTMLDivElement>) => {
    e.stopPropagation();
  };
  const modal = (
    <Suspense fallback={<Loader />}>
      <ModalOuter
        aria-modal
        aria-label={title}
        role="dialog"
        tabIndex={-1}
        onClick={onDismiss}
      >
        <ModalDialog size={size} role="document">
          <ModalContent onClick={doNotDismiss}>
            <Card padding="lg">{children}</Card>
          </ModalContent>
        </ModalDialog>
      </ModalOuter>

      <Overlay />
    </Suspense>
  );

  useEffect(() => {
    const keyboardEvent = (e: KeyboardEvent) => {
      if (e.key === 'Escape') {
        onDismiss();
      }
    };

    document.body.appendChild(output);

    document.addEventListener('keydown', keyboardEvent, false);

    return () => {
      document.body.removeChild(output);

      document.body.style.removeProperty('overflow');

      document.removeEventListener('keydown', keyboardEvent, false);
    };
  }, [output, onDismiss]);

  useEffect(() => {
    if (isOpen) {
      document.body.style.overflow = 'hidden';
    }
  }, [isOpen]);

  if (isOpen) {
    return createPortal(modal, output);
  }

  return null;
};

export default Modal;
