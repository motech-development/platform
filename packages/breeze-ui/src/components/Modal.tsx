import { Dialog, Transition } from '@headlessui/react';
import {
  ElementType,
  Fragment,
  isValidElement,
  MutableRefObject,
  ReactNode,
} from 'react';
import { Box, PolymorphicComponentProps } from 'react-polymorphic-box';
import Overlay from './Overlay';
import Typography from './Typography';

interface IModalProps {
  content: ReactNode | string;
  footer: ReactNode;
  initialFocusRef?: MutableRefObject<HTMLButtonElement | null>;
  isOpen: boolean;
  title: string;
  onDismiss: () => void;
}

export type TModalProps<E extends ElementType> = PolymorphicComponentProps<
  E,
  IModalProps
>;

const defaultElement = 'div';

const Modal = <E extends ElementType = typeof defaultElement>({
  content,
  footer,
  initialFocusRef,
  isOpen,
  onDismiss,
  title,
  ...rest
}: TModalProps<E>) => (
  <Transition.Root show={isOpen} as={Fragment}>
    <Dialog
      as="div"
      className="fixed z-10 inset-0 overflow-y-auto"
      initialFocus={initialFocusRef}
      onClose={onDismiss}
    >
      <div className="flex items-end justify-center min-h-screen pt-4 px-4 pb-20 text-center sm:block sm:p-0">
        <Transition.Child
          as={Fragment}
          enter="ease-out duration-300"
          enterFrom="opacity-0"
          enterTo="opacity-100"
          leave="ease-in duration-200"
          leaveFrom="opacity-100"
          leaveTo="opacity-0"
        >
          <Dialog.Overlay as={Overlay} />
        </Transition.Child>

        <span
          className="hidden sm:inline-block sm:align-middle sm:h-screen"
          aria-hidden="true"
        >
          &#8203;
        </span>

        <Transition.Child
          as={Fragment}
          enter="ease-out duration-300"
          enterFrom="opacity-0 translate-y-4 sm:translate-y-0 sm:scale-95"
          enterTo="opacity-100 translate-y-0 sm:scale-100"
          leave="ease-in duration-200"
          leaveFrom="opacity-100 translate-y-0 sm:scale-100"
          leaveTo="opacity-0 translate-y-4 sm:translate-y-0 sm:scale-95"
        >
          <Box
            as={defaultElement}
            className="inline-block align-bottom bg-gray-100 px-4 pt-5 pb-4 text-left overflow-hidden shadow-xl transform transition-all sm:my-8 sm:align-middle sm:max-w-lg sm:w-full sm:p-6"
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...rest}
          >
            <div>
              <div className="text-center">
                <Dialog.Title
                  rule
                  as={Typography}
                  align="center"
                  component="h3"
                  variant="h3"
                  margin="lg"
                >
                  {title}
                </Dialog.Title>

                <div className="mt-2">
                  {isValidElement(content) ? (
                    content
                  ) : (
                    <Typography
                      className="text-sm text-gray-500"
                      component="p"
                      variant="p"
                      margin="none"
                      align="center"
                    >
                      {content}
                    </Typography>
                  )}
                </div>
              </div>
            </div>
            <div className="mt-6 grid grid-cols-2 gap-3 grid-flow-row-dense">
              {footer}
            </div>
          </Box>
        </Transition.Child>
      </div>
    </Dialog>
  </Transition.Root>
);

export default Modal;
