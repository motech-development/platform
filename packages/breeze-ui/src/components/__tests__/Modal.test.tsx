import { render, RenderResult, waitFor } from '@testing-library/react';
import Modal from '../Modal';

describe('Modal', () => {
  let onDismiss: jest.Mock;

  beforeEach(() => {
    onDismiss = jest.fn();
  });

  it('should not show a modal when isOpen is false', () => {
    const { queryByTestId } = render(
      <Modal
        content={<p data-testid="content">Content</p>}
        footer={<button type="button">Close</button>}
        isOpen={false}
        title="Title"
        onDismiss={onDismiss}
      />,
    );

    expect(queryByTestId('content')).not.toBeInTheDocument();
  });

  describe('when open', () => {
    let component: RenderResult;

    beforeEach(() => {
      component = render(
        <Modal
          isOpen
          content={<p data-testid="content">Content</p>}
          footer={<button type="button">Close</button>}
          title="Title"
          onDismiss={onDismiss}
        />,
      );
    });

    it('should show content', async () => {
      const { getByRole } = component;

      const dialog = await waitFor(() => getByRole('dialog'));

      expect(dialog).toMatchSnapshot();
    });
  });
});
