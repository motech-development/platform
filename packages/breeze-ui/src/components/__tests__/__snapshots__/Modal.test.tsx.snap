// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal when open should show content 1`] = `
<div
  aria-labelledby="headlessui-dialog-title-6"
  aria-modal="true"
  class="fixed z-10 inset-0 overflow-y-auto"
  id="headlessui-dialog-1"
  role="dialog"
>
  <div
    class="flex items-end justify-center min-h-screen pt-4 px-4 pb-20 text-center sm:block sm:p-0"
  >
    <div
      class="fixed inset-0 bg-gray-500 bg-opacity-75 transition-opacity"
    />
    <span
      aria-hidden="true"
      class="hidden sm:inline-block sm:align-middle sm:h-screen"
    >
      â€‹
    </span>
    <div
      class="inline-block align-bottom bg-gray-100 px-4 pt-5 pb-4 text-left overflow-hidden shadow-xl transform transition-all sm:my-8 sm:align-middle sm:max-w-lg sm:w-full sm:p-6"
    >
      <div>
        <div
          class="text-center"
        >
          <div>
            <h3
              class="mb-2.5 text-center font-display text-3xl font-semibold"
              id="headlessui-dialog-title-6"
            >
              Title
            </h3>
            <hr
              class="w-12 border-t-4 border-red-600 shadow-md m-auto mb-2.5"
            />
          </div>
          <div
            class="mt-2"
          >
            <p
              data-testid="content"
            >
              Content
            </p>
          </div>
        </div>
      </div>
      <div
        class="mt-6 grid grid-cols-2 gap-3 grid-flow-row-dense"
      >
        <button
          tabindex="0"
          type="button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;
